File: Python/queue.py
from typing import TypeVar
from webbrowser import get

T = TypeVar("T")

type QueueNode[T] = {
    val: T,
    next: QueueNode[T] | None,
}


class LinkedListQueue[T]:
    def __init__(self) -> None:
        self.front: QueueNode[T] | None = None
        self.rear: QueueNode[T] | None = None
        self.queueSize: int = 0

    def enqueue(self, value: T) -> None:
        node: QueueNode[T] = { val: value, next: None }
        if not self.front:
            self.front = node
            self.rear = node
        else:
            self.rear.next = node
            self.rear = node
        self.queueSize += 1
    
    def dequeue(self) -> T:
        if not self.front:
            raise Exception("Queue is empty")
        
        node = self.front
        self.front = self.front.next
        self.queueSize -= 1
        return node.val
    
    def peek(self) -> T:
        if not self.front:
            raise Exception("Queue is empty")
        return self.front.val
    
    def toArray(self) -> list[T]:
        arr = []
        currentNode = self.front
        while currentNode:
            arr.append(currentNode.val)
            currentNode = currentNode.next
        return arr

class ArrayQueue[T]:
    def __init__(self, capacity: int) -> None:
        self.queue: list[T] = [None] * capacity
        self.front: int = 0
        self.queueSize: int = 0

    @property
    def size(self) -> int:
        return self.queueSize
    
    @property
    def capacity(self) -> int:
        return len(self.queue)
    
    def isEmpty(self) -> bool:
        return self.size == 0
    
    def enqueue(self, value: T) -> None:
        if self.size == self.capacity:
            raise Exception("Queue is full")
        rear = (self.front + self.queueSize) % self.capacity
        self.queue[rear] = value
        self.queueSize += 1
    
    def dequeue(self) -> T:
        if self.size == 0:
            raise Exception("Queue is empty")
        front = self.queue[self.front]
        self.front = (self.front + 1) % self.capacity
        self.queueSize -= 1
        return front
    
    def peek(self) -> T:
        if self.size == 0:
            raise Exception("Queue is empty")
        return self.queue[self.front]
    
    def toArray(self) -> list[T]:
        arr = []
        for i in range(self.queueSize):
            arr.append(self.queue[(self.front + i) % self.capacity])
        
        return arr

if __name__ == "__main__":
    queue = ArrayQueue[int](10)
    for i in range(10):
        queue.enqueue(i)
    print(queue.toArray())
    print(queue.dequeue())
    print(queue.toArray())
    

File: TS/src/queue.ts
type QueueNode<T> = {
  val: T;
  next: QueueNode<T> | null;
};

class LinkedListQueue<T> {
  constructor(
    private front: QueueNode<T> | null = null,
    private rear: QueueNode<T> | null = null,
    private queueSize: number = 0
  ) {}

  get size(): number {
    return this.queueSize;
  }

  isEmpty(): boolean {
    return this.size === 0;
  }

  enqueue(value: T): void {
    const node: QueueNode<T> = { val: value, next: null };
    if (!this.front) {
      this.front = node;
      this.rear = node;
    } else {
      this.rear!.next = node;
      this.rear = node;
    }
    this.queueSize++;
  }

  dequeue(): T {
    if (!this.front) throw new Error("Queue is empty");
    const node = this.front;
    this.front = this.front.next;
    this.queueSize--;
    return node.val;
  }

  peek(): T {
    if (this.size === 0) throw new Error("Queue is empty");
    return this.front!.val;
  }

  toArray(): T[] {
    let currentNode = this.front;
    const arr: T[] = new SeqList<T>(this.queueSize);
    for (let i = 0; i < this.queueSize; i++) {
      arr[i] = currentNode!.val;
      currentNode = currentNode!.next;
    }
    return arr;
  }
}

class ArrayQueue<T> {
  constructor(
    private queue: T[] = [],
    private front: number = 0,
    private queueSize: number = 0
  ) {}

  get capacity(): number {
    return this.queue.length;
  }

  get size(): number {
    return this.queueSize;
  }

  isEmpty(): boolean {
    return this.size === 0;
  }

  enqueue(value: T): void {
    if (this.size === this.capacity) throw new Error("Queue is full");
    const rear = (this.front + this.queueSize) % this.capacity;
    this.queue[rear] = value;
    this.queueSize++;
  }

  dequeue(): T {
    const front = this.peek();
    this.front = (this.front + 1) % this.capacity;
    this.queueSize--;
    return front;
  }

  peek(): T {
    if (this.size === 0) throw new Error("Queue is empty");
    return this.queue[this.front];
  }

  toArray(): T[] {
    const arr = new SeqList<T>(this.queueSize);
    for (let i = 0, j = this.front; i < this.queueSize; i++, j++) {
      arr[i] = this.queue[j % this.capacity];
    }
    return arr;
  }
}
